db/dcrpg/internal/txstmts.go

```diff
@@ -130,8 +130,17 @@ const (
 	SelectRegularTxByHash = `SELECT id, block_hash, block_index FROM transactions WHERE tx_hash = $1 and tree=0;`
 	SelectStakeTxByHash   = `SELECT id, block_hash, block_index FROM transactions WHERE tx_hash = $1 and tree=1;`
 
+	SelectAllTicketsBlockTime = `SELECT block_height, block_time FROM transactions WHERE tx_type = 1;`
+
+	SelectTicketsByType = `SELECT
+		SUM(CASE WHEN num_vout = 3 THEN 1 ELSE 0 END) as solo,
+		SUM(CASE WHEN num_vout = 5 THEN 1 ELSE 0 END) as pooled,
+		SUM(CASE WHEN num_vout > 5 THEN 1 ELSE 0 END) as tixsplit
+		FROM transactions WHERE tx_hash = ANY(SELECT tx_hash FROM tickets
+		WHERE pool_status=0 AND is_mainchain = TRUE) AND tx_type=1;`
```

Comment:
Of course we might pay a non-trivial penalty on transaction inserts.

Reply:
Some more thoughts on this query:

- The `AND tx_type=1` query is not necessary since the tx hash is coming from the tickets table.  Removing that cuts the query time more than in half.
- The tickets table has the corresponding row id (the primary key) in the transactions table, so you can use that instead of tx_hash.  That cuts the query down to about 95 ms.
- A `JOIN` instead of the `WHERE tx_hash = ANY(SELECT` might be better.
- The `SUM(CASE` stuff can be avoided with `GROUP BY num_vout` with just a little aggregation for the >5 cases in Go.

This gives:

```
dcrdata=# SELECT num_vout, count(*) FROM transactions JOIN tickets ON transactions.id=purchase_tx_db_id WHERE pool_status=0 AND tickets.is_mainchain = TRUE GROUP BY num_vout;
 num_vout | count 
----------+-------
        3 | 21469
        5 | 20765
        7 |     1
        9 |     3
       11 |     2
       13 |     2
       15 |     2
       17 |     1
       19 |     2
       25 |     1
(10 rows)
```

With the first two changes, it's down to <100ms, but the other two change seem to help readability to me.

Prettied-up, the query is:

```sql
SELECT num_vout,
       count(*)
FROM transactions
JOIN tickets ON transactions.id=purchase_tx_db_id
WHERE pool_status=0
  AND tickets.is_mainchain = TRUE
GROUP BY num_vout;
```

Reply:
Or sticking with the `SUM(CASE` approach, and applying the first 3 changes:

```
dcrdata=# SELECT SUM(CASE WHEN num_vout = 3 THEN 1 ELSE 0 END) as solo, SUM(CASE WHEN num_vout = 5 THEN 1 ELSE 0 END) as pooled, SUM(CASE WHEN num_vout > 5 THEN 1 ELSE 0 END) as tixsplit FROM transactions JOIN tickets ON transactions.id=purchase_tx_db_id WHERE pool_status=0 AND tickets.is_mainchain = TRUE;
 solo  | pooled | tixsplit 
-------+--------+----------
 21469 |  20765 |       14
(1 row)
```

Reply:
More naturally, using the `width_bucket` aggregate and keeping the table output format:

```sql
SELECT width_bucket(num_vout, array[3, 5, 6]) as ticket_bucket,
       count(*)
FROM transactions
JOIN tickets ON transactions.id=purchase_tx_db_id
WHERE pool_status=0                         -- live tickets
  AND tickets.is_mainchain = TRUE
GROUP BY ticket_bucket;
```
 Output:

```
 ticket_bucket | count 
---------------+-------
             1 | 21469
             2 | 20765
             3 |    14
(3 rows)
```

Reply:
ok. Noted

Reply:
For sure this new query version is faster,
```
dcrdata=# explain analyze SELECT width_bucket(num_vout, array[3, 5, 6]) as ticket_bucket,  count(*) FROM transactions JOIN tickets ON transactions.id=purchase_tx_db_id WHERE pool_status=0   AND tickets.is_mainchain = TRUE GROUP BY ticket_bucket;
                                                                                   QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=157614.16..157615.90 rows=87 width=12) (actual time=357.169..357.171 rows=3 loops=1)
   Group Key: (width_bucket(transactions.num_vout, '{3,5,6}'::integer[]))
   ->  Sort  (cost=157614.16..157614.37 rows=87 width=12) (actual time=357.166..357.167 rows=6 loops=1)
         Sort Key: (width_bucket(transactions.num_vout, '{3,5,6}'::integer[]))
         Sort Method: quicksort  Memory: 25kB
         ->  Gather  (cost=157601.57..157611.35 rows=87 width=12) (actual time=357.143..362.777 rows=6 loops=1)
               Workers Planned: 1
               Workers Launched: 1
               ->  Partial HashAggregate  (cost=156601.57..156602.65 rows=87 width=12) (actual time=348.596..348.597 rows=3 loops=2)
                     Group Key: width_bucket(transactions.num_vout, '{3,5,6}'::integer[])
                     ->  Nested Loop  (cost=0.86..156484.10 rows=23494 width=4) (actual time=0.986..342.367 rows=21182 loops=2)
                           ->  Parallel Index Scan using pool_status_index on tickets  (cost=0.43..28881.31 rows=23494 width=8) (actual time=0.461..197.584 rows=21182 loops=2)
                                 Index Cond: (pool_status = 0)
                                 Filter: is_mainchain
                           ->  Index Scan using transactions_pkey on transactions  (cost=0.43..5.43 rows=1 width=12) (actual time=0.006..0.006 rows=1 loops=42364)
                                 Index Cond: (id = tickets.purchase_tx_db_id)
 Planning time: 2.520 ms
 Execution time: 362.918 ms
(18 rows)
```

Compared To
```
dcrdata=# explain analyze SELECT SUM(CASE WHEN num_vout = 3 THEN 1 ELSE 0 END) as solo, SUM(CASE WHEN num_vout = 5 THEN 1 ELSE 0 END) as pooled, SUM(CASE WHEN num_vout > 5 THEN 1 ELSE 0 END) as tixsplit FROM transactions WHERE tx_hash = ANY(SELECT tx_hash FROM tickets WHERE pool_status=0 AND is_mainchain = TRUE) AND tx_type=1;
                                                                               QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=285825.65..285825.66 rows=1 width=24) (actual time=7391.852..7391.852 rows=1 loops=1)
   ->  Gather  (cost=285825.43..285825.64 rows=2 width=24) (actual time=7374.604..7399.996 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial Aggregate  (cost=284825.43..284825.44 rows=1 width=24) (actual time=7375.099..7375.099 rows=1 loops=3)
               ->  Hash Semi Join  (cost=30014.00..284753.26 rows=4811 width=4) (actual time=6655.908..7369.732 rows=14121 loops=3)
                     Hash Cond: (transactions.tx_hash = tickets.tx_hash)
                     ->  Parallel Seq Scan on transactions  (cost=0.00..239126.91 rows=578984 width=69) (actual time=0.338..6659.978 rows=461408 loops=3)
                           Filter: (tx_type = 1)
                           Rows Removed by Filter: 1140639
                     ->  Hash  (cost=29045.76..29045.76 rows=39939 width=65) (actual time=350.220..350.220 rows=42364 loops=3)
                           Buckets: 65536  Batches: 2  Memory Usage: 2514kB
                           ->  Index Scan using pool_status_index on tickets  (cost=0.43..29045.76 rows=39939 width=65) (actual time=0.353..325.339 rows=42364 loops=3)
                                 Index Cond: (pool_status = 0)
                                 Filter: is_mainchain
 Planning time: 7.588 ms
 Execution time: 7400.215 ms
(17 rows)
```